apps/sms/js/startup.js

```diff
@@ -112,17 +112,19 @@ var Startup = {
       MessageManager.init();
       Navigation.init();
 
-      // If target panel is different from the default one, let's load remaining
-      // scripts as soon as possible, otherwise we can wait until first page of
-      // threads is loaded and rendered.
-      var panelName = Navigation.getPanelName();
-      if (panelName && panelName !== 'thread-list') {
-        // Initialize default panel only after target panel is ready.
+      var panelName = Navigation.getPanelName() || Navigation.defaultPanel;
+
+      // If initial panel is default one and app isn't run from notification,
+      // then just navigate to it, otherwise we can delay default panel
+      // initialization until we navigate to requested non-default panel.
+      if (panelName === Navigation.defaultPanel &&
+        !navigator.mozHasPendingMessage('notification')) {
+        Navigation.toPanel(panelName);
```

Comment:
> grrr I'd like to use mozHasPendingMessage without an argument but it's not possible :( For example we probably want to do the same with sms-received so that we get the notification earlier... Not as important though but this is not a really good code :/

I don't like mozHasPendingMessage either, see https://bugzilla.mozilla.org/show_bug.cgi?id=1162899 that was bothering me too much yesterday :(

> We want to move to the defaultPanel only when:
* there is no message (means no hash (getPanelName() returns the empty string) and no pending message for 'notification')
* no other panel has been requested

Mmm, what the difference between " there is no message (means no hash (getPanelName() returns the empty string)" and "no other panel has been requested"?

> Here in this code I don't like that we use a lot of functions from Navigation, I feel like some or most of this code should be in Navigation.

Yes, too much, agree :)

> So I'd keep the existing toPanelFromHash (but I'd still remove the call from Navigation.init) and do something like:

```javascript
if (Navigation.isDefaultPanel() && !navigator.mozHasPendingMessage('notification')) {
  Navigation.toDefaultPanel(); // here this could return a promise, and we could put the setReady/fullyLoaded mark as handler as this promise (same for toPanelFromHash below)
  initializeDefaultPanel(...); // maybe the content of this could be moved to `InboxView.beforeEnter()` (without returning the promise) ?
} else {
  Navigation.toPanelFromHash(); // modified to not move to default panel if there is no hash
  Navigation.on('navigated', ...); // here the handler could be simply InboxView.beforeEnter()
}
```

Honestly, I don't see any value in ```toPanelFromHash``` right now, it's a bit confusing when we have "path" that never used - for master it's not used currently and will likely be modified for ng, for v2.2 it won't likely be used ever, why do you want to still keep it? We subscribe to ```Navigation.on('navigated')``` because we know that later navigation will happen (in activity handler), if it doesn't - its a different story that we don't handle correctly right now anyway :)

So you'd like to extract ```!panel || panelName === Navigation.defaultPanel``` to ```Navigation.isDefaultPanel``` and ```Navigation.toPanel(Navigation.defaultPanel)``` to ```Navigation.toDefaultPanel``` - I think this makes sense, do you feel that ```Navigation.defaultPanel``` should be "private"?

> initializeDefaultPanel(...) // maybe the content of this could be moved to `InboxView.beforeEnter()` (without returning the promise) ?

That's what I was thinking from the "beginning", we should do it eventually, but I'm not sure if it's safe for v2.2 patch though - I'll experiment. But if we don't return a promise, how do you mark ```markFullyLoaded``` before threads are fully rendered in the "extended" code snippet?

> Navigation.toPanelFromHash(); // modified to not move to default panel if there is no hash

As I said above, we don't use it - why do we need this? I'd rather add it only when\if we need it :)

```
Navigation.toPanelFromHash().then(
  function onLoaded() {
    Navigation.on('navigated', InboxView.beforeEnter.bind(InboxView));
```

Mmm, I think ```navigated``` is fired before ```onLoaded``` is called, so ```InboxView.beforeEnter``` will fire at some undetermined moment later when user navigates to somewhere else, e.g. to InboxView :)

> anyway in your current code we never move to the panel if it's not default :)

Mm, activity handler moves to the panel, since we don't have other cases for non-default hashes other than ones handled by activity handler. And if something fails in activity handler and navigation doesn't happen then we stuck even on current master :)

So to summarize: 
* I agree that smth like ```Navigation.method(Navigation.property)``` feels weird and it's better to have ```Navigation.methodToDealWithProperty()``` - I'll add them;
* I agree that it would be great to have ```initDefaultPanel``` inside of Inbox view, though I have two concerns: 
 * it maybe too risky to do for v2.2;
 * InboxView is special - it should render threads even if we don't navigate to it and "Inbox.beforeEnter" feels "internal/protected" (don't have such in js, but I mean it should be accessible only by classes from the same "area/namespace" like Navigation) and accepts parameters type that only Navigation knows - I'll think and experiment since we'll should do smth like this in NG anyway.
* I still don't think we need ```toPanelFromHash()``` - it's confusing since it's not used, especially for v2.2. We have either ```Navigation.toDefaultPanel``` called from startup or ```Navigation.toPanel(specific_panel)``` called from activity handler.